### DISCUSSION TITLE
freeplane2md - a feature rich Freeplane to Markdown conversion

### DISCUSSION THREAD

**Discussion by OliverCi** [freeplane/freeplane#1200] [Show and tell]
I am a regular Freeplane user and apply it for many purposes. Often, I need the resulting mind map in Markdown format as well. Unfortunately, the built in export functionality so far can produce only headers. Some other scripts did also not meet my needs. Thus, I wrote [freeplane2md](https://github.com/OliverCi/freeplane2md), a command line tool and scriptable Python function to convert mind maps in Freeplane format into Markdown. So far, it has the folllowing features:

- Freeplane nodes are converted to Markdown headers down to a configurable level and to list items below that level.
- Freeplane links are converted to Markdown links:
  - Links to URLs or E-Mail addresses are converted to automatic links.
  - For local links within the document, custom IDs are generated from the Freeplane node IDs.
  - Links to local Markdown files are converted to `[[WikiLinks]]` as supported by most tools.
- Freeplane connections between nodes are converted to document local links.
- Checkbox icons are converted to Markdown extended syntax checkboxes (`- [ ]`, `- [x]`).
- Frequently used icons are converted to emojis (currently only in shortcode syntax), as understood by most Markdown tools.
- Some limited support for HTML embedded in nodes.

Maybe it is helpful to further Freeplane users. Let me know your comments.

**Comment by Illegit** [comment]
Hi OliverCi
I installed everything but I got the following error: python freeplane2md.py XXXXX.mm
Traceback (most recent call last):
  File "freeplane2md.py", line 59, in <module>
    from docopt import docopt
ModuleNotFoundError: No module named 'docopt'

Any suggestion?

**Reply by sifran-github** [reply]
look at the installation part at https://github.com/OliverCi/freeplane2md#installation
you need to install these required modules to run the script
for docopt you should run: pip install docopt

install the other modules as well, and it should work.

**Comment by Illegit** [comment]
I confiirm I've installed everything

e.g.: 
pip install docopt
Requirement already satisfied: docopt in c:\users\xxx\appdata\local\programs\python\python311\lib\site-packages (0.6.2)

pip install validators
Requirement already satisfied: validators in c:\users\xxx\appdata\local\programs\python\python311\lib\site-packages (0.20.0)
Requirement already satisfied: decorator>=3.4.0 in 
c:\users\xxx\appdata\local\programs\python\python311\lib\site-packages (from validators) (5.1.1)

pip install python-dateutil
Requirement already satisfied: python-dateutil in c:\users\xxx\appdata\local\programs\python\python311\lib\site-packages (2.8.2)
Requirement already satisfied: six>=1.5 in c:\users\xxx\appdata\local\programs\python\python311\lib\site-packages (from python-dateutil) (1.16.0)

but I get the same error....?
this is my command line:

C:\Users\xxx\OneDrive\@Data\myData\Maps>python freeplane2md.py Beatles.mm

> this is the system result:

Traceback (most recent call last):
  File "freeplane2md.py", line 59, in <module>
    from docopt import docopt
ModuleNotFoundError: No module named 'docopt'

So how to use freeplane2md.py?

**Reply by OliverCi** [reply]
Hi Illegit, 
maybe `C:\Users\xxx\OneDrive@Data\myData\Maps>python freeplane2md.py Beatles.mm` is executing under a different Python version than pip in your installation,, see e.g. <https://stackoverflow.com/questions/49880271/modulenotfounderror-no-module-named-docopt>. (You can check with `python --version`).

**Reply by Illegit** [reply]
Hi OliverCi, I've already investigated, but I see some Python 3.8 even but when I read the installed python app there is only 3.11... so I do not know what to do. Probably I uninstall python and then reinstall it...

**Reply by OliverCi** [reply]
Yes, this migth work, if you also add Python to your PATH during install. Before that, you can also try

- if selecting Python 3.11 works, e.g. `py -3.11 freeplane2md.py Beatles.mm`, or
- install the modules for the running Python (3.8?), e.g. `python -m pip install docopt` etc. (see for example <https://stackoverflow.com/questions/2812520/dealing-with-multiple-python-versions-and-pip>).

**Reply by Illegit** [reply]
This last command (py -3.11 freeplane2md.py Beatles.mm) is working! Now the problem is that it is not able to recognize if there is a space in the file title, it takes the part separated by the space as an option for the script. Still, the script is simple without identifying icons attributes images in details or nodes, but keeping exactly the links. And it is even fast. Thank you. I think I can use it for some simple/mid-complex maps.

**Comment by Illegit** [comment]
Currently, I use the Markdown Helper of EdoFro to create documentation, it is nice but a little complicate-slow, I'm looking for a faster way.

**Reply by macmarrum** [reply]
You can have a look at a scripting solution, like this one:
- https://github.com/freeplane/freeplane/issues/333

**Comment by Illegit** [comment]
Thanks, macmarrum, I tried it but the result is fast but not quite satisfying because it misses the structure of nodes. In complex maps, I couldn't understand the structure of reading in MD, and I'm not so skilled in MD to be able to modify it effectively. E.g.: I tried to add more than 4 levels but already at the 5th it became the writing with: ##### etc.
